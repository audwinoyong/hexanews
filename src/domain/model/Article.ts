/**
 * Model for the article.
 */
export type Article = {
  title: string,
  description: string
  category?: string
  authorId?: string
}