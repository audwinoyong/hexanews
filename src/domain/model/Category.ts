/**
 * Model for category object in our application.
 */

export type Category = {
  _id?: string,
  title: string,
  description: string,
}