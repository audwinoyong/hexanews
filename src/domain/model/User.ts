/**
 * Model for the user.
 */
export type User = {
  _id: string,
  email: string,
  name: string,
  description: string,
  createdAt: string,
}