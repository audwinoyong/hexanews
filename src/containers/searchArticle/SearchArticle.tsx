/**
 * Display search article container for searching any articles based on the query
 */

import { CardMedia } from '@material-ui/core'
import { connect } from 'react-redux'
import { Dispatch } from 'redux'
import { isEmpty, map, filter, find, head, startCase } from 'lodash'
import { translate } from 'react-i18next'
import * as moment from 'moment'
import * as React from 'react'
import Avatar from '@material-ui/core/Avatar'
import Card from '@material-ui/core/Card'
import CardContent from '@material-ui/core/CardContent'
import CardHeader from '@material-ui/core/CardHeader'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import Typography from '@material-ui/core/Typography'

import { Article } from '../../domain/model/Article'
import { DATE_FORMAT } from '../../constants/config'
import { User } from '../../domain/model/User'
import i18n from '../../i18n'
import placeholder from '../../assets/placeholder.png'
import selector, { StateProps } from './selector'
import styles from './styles'
import Utils from '../../utils'

/**
 * All props required by the container
 */
export type Props = {
  dispatch: Dispatch<any>
} & StateProps

/**
 * All state required by the container
 */
export interface ComponentState {
  query: string
}

export class SearchArticle extends React.Component<Props, ComponentState> {

  /**
   * Main constructor for Search Article container and initialize state
   * @param props props required in the components
   */
  constructor(props: Props) {
    super(props)
    this.state = {
      query: ''
    }
  }

  /**
   * Function that get the filtered articles from the query entered
   */
  getFilteredArticles = () => {
    const { articles } = this.props
    const query = this.state.query.trim().toLowerCase()
    return filter(articles, (article: any) => article.title.toLowerCase().indexOf(query) !== -1)
  }

  /**
   * Function that handle onChange text in query textfield
   * @param event contain the query value from textField
   */
  handleQueryChange = (event: any) => {
    this.setState({ query: event.target.value })
  }

  /**
   * Function that render an avatar of the author's initals
   * @param authorName the name of the author
   */
  renderAvatar = (authorName: string) => {
    const initials = head(startCase(authorName))
    return <Avatar style={styles.avatar}>{initials}</Avatar>
  }

  /**
   * Function that render all the filtered articles
   */
  renderArticles = (filteredArticles: Article[]) => {
    const { users } = this.props

    if (isEmpty(filteredArticles)) {
      return (
        <Grid style={styles.emptyArticle}>
          <Typography variant='subheading'>
            {i18n.t('searchArticle.noArticlesFound')}
          </Typography>
        </Grid>
      )
    }

    return map(filteredArticles, (article: Article) => {
      const author = find(users, (user: User) => user._id === article.authorId)
      const authorName = author ? author.name : ''
      const articleImage = Utils.getFeaturedImage(article) ? Utils.getFeaturedImage(article) : placeholder
      const sanitizedDescription = article.description.replace(/<(?:.|\n)*?>/gm, '')
      const articleDescription = sanitizedDescription.length > 128 ? `${sanitizedDescription.substring(0, 128)}...` : sanitizedDescription
      return (
        <Card style={styles.card} key={article._id}>
          <CardHeader
            avatar={this.renderAvatar(authorName)}
            title={authorName}
            subheader={moment(article.createdAt).format(DATE_FORMAT)}
          />
          <CardMedia
            style={styles.media}
            image={articleImage}
          />
          <CardContent>
            <Typography gutterBottom variant='headline' component='h2'>
              {article.title}
            </Typography>
            <Typography component='p'>
              {articleDescription}
            </Typography>
          </CardContent>
        </Card>
      )
    })
  }

  /**
   * Render Search Article container
   */
  public render() {
    const query = this.state.query
    const filteredArticles = this.getFilteredArticles()
    return (
      <Grid style={styles.container}>
        <Grid container>
          <Grid item md={3} xs={1}/>
          <Grid item md={6} xs={10}>
            <TextField
              id='query'
              label={i18n.t('searchArticle.search')}
              value={query}
              onChange={this.handleQueryChange}
              style={styles.textField}
              margin='normal'
            />
          </Grid>
          <Grid item md={3} xs={1}/>
        </Grid>
        <Grid container spacing={24}>
          <Grid item md={3} xs={1}/>
          <Grid item md={6} xs={10} style={styles.articleContainer}>
            {this.renderArticles(filteredArticles)}
          </Grid>
          <Grid item md={3} xs={1}/>
        </Grid>
      </Grid>
    )
  }
}

const ConnectedSearchArticle = connect(selector)(SearchArticle)

export default translate('translations')(ConnectedSearchArticle)