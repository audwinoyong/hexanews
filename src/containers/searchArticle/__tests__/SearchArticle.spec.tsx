import * as React from 'react'
import * as renderer from 'react-test-renderer'
import * as Enzyme from 'enzyme'
import { shallow } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

import { SearchArticle, Props, ComponentState } from '../SearchArticle'
import { Article } from '../../../domain/model/Article'

Enzyme.configure({ adapter: new Adapter() })

const mockArticles: Article[] = [
  {
    _id: '1',
    title: 'abc',
    description: 'test',
    category: 'test',
    authorId: 'test',
  },
  {
    _id: '2',
    title: 'def',
    description: 'tester',
    category: 'tester',
    authorId: 'tester',
  }
]

describe('Search Article', () => {
  it('renders Search Articles with zero articles', () => {
    const component = renderer.create(
      <SearchArticle
        articles={[]}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders Search Articles with two articles', () => {
    const component = renderer.create(
      <SearchArticle
        articles={mockArticles}
      />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('handle handleQueryChange action correctly', () => {
    const component = (
      <SearchArticle
        articles={mockArticles}
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: SearchArticle = layout.instance() as SearchArticle

    const queryState = layout.state().query
    expect(queryState).toBe('')

    const event = {
      target: {
        value: 'asdf'
      }
    }
    instance.handleQueryChange(event)
    layout.update()

    const updatedQueryState = layout.state().query
    expect(updatedQueryState).toBe('asdf')
  })

  it('handle getFilteredArticles action correctly', () => {
    const component = (
      <SearchArticle
        articles={mockArticles}
      />
    )

    const layout = shallow<Props, ComponentState>(component)
    const instance: SearchArticle = layout.instance() as SearchArticle

    const initialArticles = instance.getFilteredArticles()
    expect(initialArticles).toEqual(mockArticles)

    layout.setState({ query: 'a' })
    layout.update()

    const updatedArticles = instance.getFilteredArticles()
    expect(updatedArticles).toEqual([mockArticles[0]])

    layout.setState({ query: 'd' })
    layout.update()

    const updatedArticles2 = instance.getFilteredArticles()
    expect(updatedArticles2).toEqual([mockArticles[1]])
  })
})