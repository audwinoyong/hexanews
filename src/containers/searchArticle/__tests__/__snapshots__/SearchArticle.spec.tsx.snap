// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Article renders Search Articles with two articles 1`] = `
<div
  style={
    Object {
      "marginTop": "4%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
    >
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-98 MuiFormControl-marginNormal-99"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-107 MuiInputLabel-root-102 MuiInputLabel-formControl-103 MuiInputLabel-animated-106"
          data-shrink={false}
          htmlFor="query"
        >
          Search
        </label>
        <div
          className="MuiInput-root-114 MuiInput-formControl-115 MuiInput-underline-118"
        >
          <input
            aria-invalid={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-122"
            defaultValue={undefined}
            disabled={false}
            id="query"
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={false}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
  >
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={
        Object {
          "marginRight": 24,
        }
      }
    >
      <div
        className="MuiPaper-root-128 MuiPaper-elevation1-131 MuiPaper-rounded-129 MuiCard-root-127"
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-155"
        >
          <div
            className="MuiCardHeader-avatar-156"
          >
            <div
              className="MuiAvatar-root-161"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            />
          </div>
          <div
            className="MuiCardHeader-content-158"
          >
            <span
              className="MuiTypography-root-164 MuiTypography-body2-172 MuiCardHeader-title-159"
            >
              
            </span>
            <span
              className="MuiTypography-root-164 MuiTypography-body2-172 MuiTypography-colorTextSecondary-187 MuiCardHeader-subheader-160"
            >
              12 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-189"
        >
          <h2
            className="MuiTypography-root-164 MuiTypography-headline-169 MuiTypography-gutterBottom-181"
          >
            abc
          </h2>
          <p
            className="MuiTypography-root-164 MuiTypography-body1-173"
          >
            test
          </p>
        </div>
      </div>
      <div
        className="MuiPaper-root-128 MuiPaper-elevation1-131 MuiPaper-rounded-129 MuiCard-root-127"
        style={
          Object {
            "boxShadow": "0 1px 4px rgba(0, 0, 0, 0.2)",
            "marginTop": 36,
          }
        }
      >
        <div
          className="MuiCardHeader-root-155"
        >
          <div
            className="MuiCardHeader-avatar-156"
          >
            <div
              className="MuiAvatar-root-161"
              style={
                Object {
                  "alignSelf": "center",
                  "backgroundColor": "green",
                  "height": 50,
                  "width": 50,
                }
              }
            />
          </div>
          <div
            className="MuiCardHeader-content-158"
          >
            <span
              className="MuiTypography-root-164 MuiTypography-body2-172 MuiCardHeader-title-159"
            >
              
            </span>
            <span
              className="MuiTypography-root-164 MuiTypography-body2-172 MuiTypography-colorTextSecondary-187 MuiCardHeader-subheader-160"
            >
              12 September 2018
            </span>
          </div>
        </div>
        <div
          className="MuiCardContent-root-189"
        >
          <h2
            className="MuiTypography-root-164 MuiTypography-headline-169 MuiTypography-gutterBottom-181"
          >
            def
          </h2>
          <p
            className="MuiTypography-root-164 MuiTypography-body1-173"
          >
            tester
          </p>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
</div>
`;

exports[`Search Article renders Search Articles with zero articles 1`] = `
<div
  style={
    Object {
      "marginTop": "4%",
      "paddingBottom": "5%",
    }
  }
>
  <div
    className="MuiGrid-container-1"
  >
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
    >
      <div
        aria-describedby={undefined}
        className="MuiFormControl-root-98 MuiFormControl-marginNormal-99"
        style={
          Object {
            "width": "100%",
          }
        }
      >
        <label
          className="MuiFormLabel-root-107 MuiInputLabel-root-102 MuiInputLabel-formControl-103 MuiInputLabel-animated-106"
          data-shrink={false}
          htmlFor="query"
        >
          Search
        </label>
        <div
          className="MuiInput-root-114 MuiInput-formControl-115 MuiInput-underline-118"
        >
          <input
            aria-invalid={false}
            autoComplete={undefined}
            autoFocus={undefined}
            className="MuiInput-input-122"
            defaultValue={undefined}
            disabled={false}
            id="query"
            name={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={undefined}
            onKeyUp={undefined}
            placeholder={undefined}
            readOnly={undefined}
            required={false}
            rows={undefined}
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
  <div
    className="MuiGrid-container-1 MuiGrid-spacing-xs-24-25"
  >
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
    <div
      className="MuiGrid-item-2 MuiGrid-grid-xs-10-39 MuiGrid-grid-md-6-63"
      style={
        Object {
          "marginRight": 24,
        }
      }
    >
      <div>
        <p>
          There is no articles found
        </p>
      </div>
    </div>
    <div
      className="MuiGrid-grid-xs-1-30 MuiGrid-grid-md-3-60"
    />
  </div>
</div>
`;
